server/channels/app/post.go

```diff
@@ -2644,3 +2601,86 @@ func (a *App) MoveThread(c request.CTX, postID string, sourceChannelID, channelI
 	c.Logger().Info(msg)
 	return nil
 }
+
+func (a *App) PermanentDeletePost(c request.CTX, postID, deleteByID string) *model.AppError {
+	post, err := a.Srv().Store().Post().GetSingle(sqlstore.RequestContextWithMaster(c), postID, true)
+	if err != nil {
+		return model.NewAppError("DeletePost", "app.post.get.app_error", nil, "", http.StatusBadRequest).Wrap(err)
+	}
+
+	if len(post.FileIds) > 0 {
+		appErr := a.PermanentDeleteFilesByPost(c, post.Id)
+		if appErr != nil && appErr.StatusCode != http.StatusNotFound {
+			return appErr
+		}
+	}
+
+	err = a.Srv().Store().Post().PermanentDeletePost(c, post.Id)
+	if err != nil {
+		return model.NewAppError("PermanentDeletePost", "app.post.permanent_delete_post.error", nil, "", http.StatusInternalServerError).Wrap(err)
+	}
+
+	appErr := a.AfterPostDeletionCleanUp(c, post, deleteByID)
+	if appErr != nil {
+		return appErr
+	}
+
+	return nil
+}
+
+func (a *App) AfterPostDeletionCleanUp(c request.CTX, post *model.Post, deleteByID string) *model.AppError {
+	channel, appErr := a.GetChannel(c, post.ChannelId)
+	if appErr != nil {
+		return appErr
+	}
+
+	if post.RootId == "" {
+		if appErr := a.DeletePersistentNotification(c, post); appErr != nil {
+			return appErr
+		}
+	}
+
+	postJSON, err := json.Marshal(post)
+	if err != nil {
+		return model.NewAppError("DeletePost", "api.marshal_error", nil, "", http.StatusInternalServerError).Wrap(err)
+	}
+
+	userMessage := model.NewWebSocketEvent(model.WebsocketEventPostDeleted, "", post.ChannelId, "", nil, "")
+	userMessage.Add("post", string(postJSON))
+	userMessage.GetBroadcast().ContainsSanitizedData = true
+	a.Publish(userMessage)
+
+	adminMessage := model.NewWebSocketEvent(model.WebsocketEventPostDeleted, "", post.ChannelId, "", nil, "")
+	adminMessage.Add("post", string(postJSON))
+	adminMessage.Add("delete_by", deleteByID)
+	adminMessage.GetBroadcast().ContainsSensitiveData = true
+	a.Publish(adminMessage)
+
+	a.Srv().Go(func() {
+		a.deleteFlaggedPosts(c, post.Id)
+	})
+
+	pluginPost := post.ForPlugin()
+	pluginContext := pluginContext(c)
+	a.Srv().Go(func() {
+		a.ch.RunMultiHook(func(hooks plugin.Hooks) bool {
+			hooks.MessageHasBeenDeleted(pluginContext, pluginPost)
```

Comment:
We kept the same client websocket action for soft deletion vs hard deletion. Do we just leave the plugin hooks as is for now? I don't think this is a breaking change because we are giving them the post struct same as before and our apis will not return these deleted posts in both cases

Reply:
The documentation of `MessageHasBeenDeleted` already makes it clear that the hook is invoked after the post was deleted from the DB (which was not true). This PR just adjusts the behaviour to match the docs. I think we are good to continue with your proposed changes. 

https://github.com/mattermost/mattermost/blob/b1fd445fcc197d2b5ad87ae0f21bdbfadebce4bf/server/public/plugin/hooks.go#L190-L195